problem-gson/src/main/java/org/zalando/problem/ProblemAdapterFactory.java

```diff
@@ -0,0 +1,150 @@
+package org.zalando.problem;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonNull;
+import com.google.gson.JsonObject;
+import com.google.gson.TypeAdapter;
+import com.google.gson.TypeAdapterFactory;
+import com.google.gson.internal.Streams;
+import com.google.gson.internal.bind.TypeAdapters;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+import javax.annotation.CheckReturnValue;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Problem {@link TypeAdapterFactory}.
+ */
+public class ProblemAdapterFactory implements TypeAdapterFactory {
+
+    private final boolean stackTraces;
+    private final Map<URI, Class<? extends Problem>> subtypes;
+    private final StatusTypeAdapter statusAdapter;
+
+    public ProblemAdapterFactory() {
+        this(Status.class);
+    }
+
+    @SafeVarargs
+    public <E extends Enum & StatusType> ProblemAdapterFactory(final Class<? extends E>... statusTypes) {
+        this(false, new StatusTypeAdapter(buildIndex(statusTypes)), new HashMap<>());
+    }
+
+    private ProblemAdapterFactory(boolean stackTraces, StatusTypeAdapter statusAdapter,
+                                  Map<URI, Class<? extends Problem>> subtypes) {
+        this.stackTraces = stackTraces;
+        this.statusAdapter = statusAdapter;
+        this.subtypes = subtypes;
+    }
+
+    @SafeVarargs
+    private static <E extends Enum & StatusType> Map<Integer, StatusType> buildIndex(
+            final Class<? extends E>... types) {
+        final Map<Integer, StatusType> index = new HashMap<>();
+
+        for (final Class<? extends E> type : types) {
+            for (final E status : type.getEnumConstants()) {
+                if (index.containsKey(status.getStatusCode())) {
+                    throw new IllegalArgumentException("Duplicate status codes are not allowed");
+                }
+                index.put(status.getStatusCode(), status);
+            }
+        }
+
+        return Collections.unmodifiableMap(index);
+    }
+
+    public ProblemAdapterFactory withStacktraces() {
+        return withStacktraces(true);
+    }
+
+    public ProblemAdapterFactory withStacktraces(final boolean stackTraces) {
+        return new ProblemAdapterFactory(stackTraces, this.statusAdapter, this.subtypes);
+    }
+
+    @CheckReturnValue
+    public <T extends Problem> ProblemAdapterFactory registerSubtype(Class<T> clazz, URI type) {
+        Objects.requireNonNull(clazz);
+        Objects.requireNonNull(type);
+        if (subtypes.containsKey(type) || subtypes.containsValue(clazz)) {
+            throw new IllegalArgumentException("class & type must be unique");
+        }
+
+        subtypes.put(type, clazz);
```

Comment:
:warning: That one mutates the map in the original instance of the factory.

You'll need to create a copy. To be on the safe side, you should wrap it into `Collections.unmodifableMap(..)` in the constructor.

Reply:
@whiskeysierra yes I should create a copy, but I think wrapping it with `Collections.unmodifiableMap(..)` would result in `UnsupportedOperationException` for next register call on returned instance. A `new HashMap<>()` in constructor should suffice.

Reply:
Copy it into a new hash map, and in the constructor wrap it into an unmodifiable map as a guard for future issues.

> would result in `UnsupportedOperationException` for next register call on returned instance.

Only if you try to modify it, which shouldn't happen. Copy is reading and fine.

Reply:
Done 👍 ⚡️